### DISCUSSION TITLE
Basic product analytics in a budibase app

### DISCUSSION THREAD

**Discussion by raziele** [Budibase/budibase#6757] [Show and tell]
One of my apps is basically a fancy form built with several steps. 

I wanted to be able to:
- Identify how many users using the app, and how much
- Understand which steps were taking and how much time it takes them
- Identify where users drop that might indicate issues with the app (e.g a field that is not well understood, too complicated fields etc.)

For this purpose, I used [PostHog](posthog.com) which is a great product analytics platform. 

![image](https://user-images.githubusercontent.com/2432693/179365774-d8a41449-b5fb-4710-8534-319d54e8cf3d.png)

Here's how I did it:
1. I signed up to PostHog.
2. Based on [PostHog API documentation](https://posthog.com/docs/api/post-only-endpoints#single-event) I created a new REST API data source in BB with a POST query that sends a single analytic event. Here's how it looks like:

![image](https://user-images.githubusercontent.com/2432693/179365906-46b88768-caa8-4057-85bf-b0ab87010328.png)

3. I then used this query whenever I wanted to track a user event. For example, here's a screenshot of how I configured a form text field to send an event when a user inputs something to the field: 

![image](https://user-images.githubusercontent.com/2432693/179366126-b32b3aaf-7ade-468c-9da4-17f7c937ebaa.png)

one can basically send an event wherever it is possible to execute a query.

A tricky part in this design was **how to identifying users**. 
Since the app has public access I'm not able to fetch the user details from BB, and there is no capability to generate a unique ID for every user who visits the app. 
To overcome this, I used [UUID Generator](https://www.uuidgenerator.net). It's a great little product that allows you to generate a unique ID whenever you call their API. I used a REST API source again to create a UUID-generating query:

![image](https://user-images.githubusercontent.com/2432693/179366374-f6d502f3-3d30-4733-ba58-c644461c4408.png)

To use it within the app, I needed a way to execute the query every time a user enters the app and save the result so I can pass it to all the PostHog event queries. I did it by placing a data provider directly under the form (according to the guidelines of a data provider, this query will be executed only once when the app is opened). 

![image](https://user-images.githubusercontent.com/2432693/179366550-e90e505f-8467-4310-9f9e-4908b21b1f35.png)

BTW, another interesting usage I found for UUID is **to allow my users to identify themselves** if they want to. 
I do it by logging the UUID with the results of the form upon submission AND presenting the UUID to the user as a submission ID - telling them they can use it for future inquires. 

Drawbacks:
- I'm not able to distinguish between several users and a user who submitted multiple times
- I'm not able to get any info on the users who visit the app (except the unique ID that was generated for them)

On both points I decided this is good enough so I didn't make any additional efforts in this direction.

**Comment by melohagan** [comment]
This is fantastic - thanks for sharing!

I'm curious what the use-case for your app was? 

For example in the past I've worked on a project that had analytics between form steps to compare how long it took users in a call centre to enter information in the new system compared the old one.

**Reply by raziele** [reply]
Thank you 😊

The exact use-case is disclosed, but I'll try to share some more context.

The app is a complex form that feeds multiple tables. I use Airtable as a database for now. 
I decided on building it with Budibase after researching several form building apps which were not flexible enough. 

In my case, we rolled out with a first version of the form with a test group and got very few submissions. 

I suspected there could be several reasons for this:
- too many fields. The form had many fields with different type.
- one or more fields that are too complicated for the target user persona. Some were in a format of a drop-down with ability to add new items. Really complex TMO from the start. 
-  not enough positive validation they are on the right track to a successful submission. I implemented some validation but a user had to wait for some time for the response of an api call.

But, I couldn't prove any of these assumptions. The only "analytics" I had was the number of submissions which was not enough.

So I tried simplifying the form, but I realized that I must have some sort of sense on what my users are doing and where they are getting stuck.  

I hope this makes sense.

**Reply by melohagan** [reply]
Okay that makes sense, thanks for the response :)
I hadn't seen anyone else add analytics to a Budibase app, so it is really cool and unique functionality!

**Comment by joebudi** [comment]
This is brilliant @raziele

**Comment by BKDragon** [comment]
With the help of @ConorWebb96 in another of my questions, it is now possible to make semi-permanent tracking for each user. This will make it possible to distinguish between PageViews and UserViews in PostHog.
See details: https://github.com/Budibase/budibase/discussions/11006
